### PR TITLE
Add XSPEC 12.11.0 support (HEAsoft 6.27)

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Add support for XSPEC 12.11.0 (released March 31 2020).

### Additive

- `XSagnslim`
- `XSzkerrbb`
- `XSbwcycl`

### Multiplicative

- `XSismdust`
- `XSolivineabs`
- `XSlogconst`
- `XSlog10con`

## Model parameter values

Updates the following parameter settings for changes made to the
model.dat file in XSPEC 12.11.0. There is no way to make these
settings depend on the XSPEC model library in use, so we chose to
use the latest values (in many cases these will be compatible
but there may be cases where the values are not valid with older
XSPEC libraries):

 - `XSagnsed`: `Htmax` parameter has changed to lie within the range
   6 to 10, and defaults to 10 (previously you could use 6 to 200
   and the default was 100).

 - `XSkerrbb`: `rflag` and `lflag` are now marked as `alwaysfrozen`
   (they have been changed to '$' parameters in model.dat)

 - `XSxscat`: `Xpos` and `Rext` soft and hard limits have increased;
   they were 0.95/0.95 and 115/119 and are now 0.99/0.999 and
   235/240 respectively

---

# Details

 - [X] support compiling against 12.11.0 (only tested on Linux)
 - [X] check this fix still works with 12.10.1 (ie get a green from Travis CI tests)
 - [X] fix failing test with 12.11.0: `sherpa/astro/ui/tests/test_astro_ui.py::sherpa/astro/ui/tests/test_astro_ui.py`  (it's a regression test that uses APEC and the APEC code has been changed in XSPEC 12.11.0) - see 7c4d4a0 (this is marked as failed because the macOS build timed out and for some reason I can't re-run just that test; you can see it passed in the next commit 62adf64)
 - [X] add support for new models in XSPEC 12.11.0
    - [X] agnslim
    - [X] zkerrbb
    - [X] ismdust
    - [X] olivineabs
    - [X] thcomp (only direct access, no Python class as waiting for #68 to be resolved)
    - [X] logconst
    - [X] log10con
    - [X] bwcycl
    (mixing models not supported)
 - [X] update documentation to include 12.11.0 in supported XSPEC versions
 - [X] update `setup.cfg` comments to include `hdsp` library component (as this is needed in recent versions of XSPEC)
 - [X] add a developer guide for how to handle an XSPEC version increase

### CODE REVIEW

**Comment by DougBurke** [comment]
About to rebase because of minor conflict with #732 (change to indentation of copyright line, so no functional changes).

**Comment by DougBurke** [comment]
@dtnguyen2 - when I worked on the C++ side for XSPEC in this, and previous PRs, I intentionally tried to keep it simple (i.e. allow repeated bits of code) rather than try and fix everything with a layer of C++ templates, as C++ templates are not my forte. I also based this on existing code in Sherpa, which is why it has a lot of old-style casting.

The XSPEC model templates also have a bunch of commented-out code, from checks that were going to be done but turned out to not-be-a-good-idea-at-the-time™

The question is whether this is a good place to fix these issues, or for them to be done in a separate PR?

You can also link to the line number of a particular commit (although it mike require a bit of work to get to the file if it isn't actually changed in that commit). For example, the latest commit in this PR has

https://github.com/sherpa/sherpa/blob/f7d4903860a7e59d77c7a2a86fbffc5ce00d0aa9/sherpa/include/sherpa/astro/xspec_extension.hh#L174

which is I assume the line 174-289 you mentioned and line 418-... is the 416-530

https://github.com/sherpa/sherpa/blob/f7d4903860a7e59d77c7a2a86fbffc5ce00d0aa9/sherpa/include/sherpa/astro/xspec_extension.hh#L418

**Comment by DougBurke** [comment]
@dtnguyen2 - for the std::transform I came up with something slightly different to yours

```
std::transform(std::begin(ear), std::end(ear), std::begin(fear),
		       [](const double val) -> FloatArrayType { return static_cast<FloatArrayType>(val); });
```

It seems to work.

**Comment by DougBurke** [comment]
@dtnguyen2 - to simply this PR I've started #859 to reduce the duplication in `xspec_extension.hh`

**Comment by DougBurke** [comment]
Rebased to address conflicts in the documentation.

**Comment by DougBurke** [comment]
@dtnguyen2 - I'd like to separate out the C++ changes from this PR (whether #859 or some other commit). Is that okay with you?

**Comment by dtnguyen2** [comment]
sounds good to me

File: docs/indices.rst [hunk 1]
```diff
@@ -35,7 +35,7 @@ Glossary
        speciality support for X-ray data formats that follow
        :term:`OGIP` standards (such as :term:`ARF` and :term:`RMF`
        files).
-       
+
    CXC
        The `Chandra X-ray Center <http://cxc.harvard.edu/>`_.
```
Comment (wmclaugh) [inline docs/indices.rst:40]: Should this be 'https' now?

Comment (DougBurke) [reply docs/indices.rst:40]: It should (although the http form will be okay for a while now). There's probably a bunch of other URLs in the docs that need updating.

Comment (DougBurke) [reply docs/indices.rst]: Updated the http cases (left fftw as is that page still seems to be http only).

**Pull request merged**